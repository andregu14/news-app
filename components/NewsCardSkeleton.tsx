import React, { useEffect, useRef } from "react";
import { View, ViewProps } from "./Themed";
import {
  StyleSheet,
  useColorScheme,
  Animated,
  useWindowDimensions,
} from "react-native";
import { LinearGradient } from "expo-linear-gradient";

const AnimatedLinearGradient = Animated.createAnimatedComponent(LinearGradient);

export default function NewsCardSkeleton({ style, ...otherProps }: ViewProps) {
  const colorScheme = useColorScheme() ?? "light";
  const { width } = useWindowDimensions();
  const shimmerAnim = useRef(new Animated.Value(-1)).current;

  const skeletonBaseColor = colorScheme === "dark" ? "#333" : "#E0E0E0";

  const shimmerGradientColors =
    colorScheme === "dark"
      ? (["transparent", "rgba(255,255,255,0.1)", "transparent"] as const)
      : (["transparent", "rgba(0,0,0,0.05)", "transparent"] as const);

  useEffect(() => {
    Animated.loop(
      Animated.timing(shimmerAnim, {
        toValue: 1,
        duration: 1200,
        useNativeDriver: true,
      })
    ).start();
  }, [shimmerAnim]);

  const translateX = shimmerAnim.interpolate({
    inputRange: [-1, 1],
    outputRange: [-width * 0.9, width * 0.9],
  });

  return (
    <View style={[styles.container, style, { backgroundColor: "transparent" }]}>
      {/* Placeholder para o TÃ­tulo */}
      <View
        style={[styles.skeletonTitle, { backgroundColor: skeletonBaseColor }]}
      />
      {/* Placeholder para o Corpo do Texto */}
      <View
        style={[
          styles.skeletonBodyLine,
          { width: "90%", backgroundColor: skeletonBaseColor },
        ]}
      />
      <View
        style={[
          styles.skeletonBodyLine,
          { width: "80%", backgroundColor: skeletonBaseColor },
        ]}
      />
      <View
        style={[
          styles.skeletonBodyLine,
          { width: "95%", backgroundColor: skeletonBaseColor },
        ]}
      />
      <View
        style={[
          styles.skeletonBodyLine,
          { width: "70%", backgroundColor: skeletonBaseColor },
        ]}
      />
      {/* Placeholder para a Imagem */}
      <View
        style={[styles.skeletonImage, { backgroundColor: skeletonBaseColor }]}
      />

      {/* Placeholder para o Footer */}
      <View
        style={[styles.skeletonFooter, { backgroundColor: skeletonBaseColor }]}
      />

      {/* Overlay com o Gradiente Animado */}
      <AnimatedLinearGradient
      colors={shimmerGradientColors}
      style={[StyleSheet.absoluteFill, { transform: [{ translateX }] }]}
      start={{ x: 0, y: 0.5 }}
      end={{ x: 1, y: 0.5 }}
    />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    marginVertical: 20,
    paddingBottom: 20,
    marginHorizontal: 20,
    overflow: "hidden",
    position: "relative",
  },
  skeletonTitle: {
    height: 18,
    width: "70%",
    borderRadius: 4,
    marginLeft: 15,
    marginBottom: 15,
  },
  skeletonBodyLine: {
    height: 14,
    borderRadius: 4,
    marginBottom: 8,
    marginHorizontal: 15,
  },
  skeletonImage: {
    width: "100%",
    height: 180,
    marginTop: 15,
    borderRadius: 4,
  },
  skeletonFooter: {
    height: 12,
    width: "50%",
    borderRadius: 4,
    marginTop: 15,
    marginLeft: 15,
  },
});
