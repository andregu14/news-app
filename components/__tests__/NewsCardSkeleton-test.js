import { render, screen } from '@testing-library/react-native';
import NewsCardSkeleton from '../NewsCardSkeleton';

describe('<NewsCardSkeleton />', () => {
    it('renders correctly', () => {
        render(<NewsCardSkeleton />);
        expect(screen.toJSON()).toMatchSnapshot();
    });
});
