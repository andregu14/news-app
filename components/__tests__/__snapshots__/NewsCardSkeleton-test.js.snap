// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewsCardSkeleton /> renders correctly 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#fff",
      },
      [
        {
          "marginHorizontal": 20,
          "marginVertical": 20,
          "overflow": "hidden",
          "paddingBottom": 20,
          "position": "relative",
        },
        undefined,
        {
          "backgroundColor": "transparent",
        },
      ],
    ]
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "#fff",
        },
        [
          {
            "borderRadius": 4,
            "height": 18,
            "marginBottom": 15,
            "marginLeft": 15,
            "width": "70%",
          },
          {
            "backgroundColor": "#E0E0E0",
          },
        ],
      ]
    }
  />
  <View
    style={
      [
        {
          "backgroundColor": "#fff",
        },
        [
          {
            "borderRadius": 4,
            "height": 14,
            "marginBottom": 8,
            "marginHorizontal": 15,
          },
          {
            "backgroundColor": "#E0E0E0",
            "width": "90%",
          },
        ],
      ]
    }
  />
  <View
    style={
      [
        {
          "backgroundColor": "#fff",
        },
        [
          {
            "borderRadius": 4,
            "height": 14,
            "marginBottom": 8,
            "marginHorizontal": 15,
          },
          {
            "backgroundColor": "#E0E0E0",
            "width": "80%",
          },
        ],
      ]
    }
  />
  <View
    style={
      [
        {
          "backgroundColor": "#fff",
        },
        [
          {
            "borderRadius": 4,
            "height": 14,
            "marginBottom": 8,
            "marginHorizontal": 15,
          },
          {
            "backgroundColor": "#E0E0E0",
            "width": "95%",
          },
        ],
      ]
    }
  />
  <View
    style={
      [
        {
          "backgroundColor": "#fff",
        },
        [
          {
            "borderRadius": 4,
            "height": 14,
            "marginBottom": 8,
            "marginHorizontal": 15,
          },
          {
            "backgroundColor": "#E0E0E0",
            "width": "70%",
          },
        ],
      ]
    }
  />
  <View
    style={
      [
        {
          "backgroundColor": "#fff",
        },
        [
          {
            "borderRadius": 4,
            "height": 180,
            "marginTop": 15,
            "width": "100%",
          },
          {
            "backgroundColor": "#E0E0E0",
          },
        ],
      ]
    }
  />
  <View
    style={
      [
        {
          "backgroundColor": "#fff",
        },
        [
          {
            "borderRadius": 4,
            "height": 12,
            "marginLeft": 15,
            "marginTop": 15,
            "width": "50%",
          },
          {
            "backgroundColor": "#E0E0E0",
          },
        ],
      ]
    }
  />
  <ViewManagerAdapter_ExpoLinearGradient
    collapsable={false}
    colors={
      [
        0,
        218103808,
        0,
      ]
    }
    endPoint={
      [
        1,
        0.5,
      ]
    }
    startPoint={
      [
        0,
        0.5,
      ]
    }
    style={
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": [
          {
            "translateX": -675,
          },
        ],
      }
    }
  />
</View>
`;
